security: TestTLSCipherRestrict failed

security.TestTLSCipherRestrict [failed](https://tanzanite.cluster.engflow.com/invocations/default/b1231e04-f605-448c-a518-f5bcd2cad19a?testReportRun=2&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc2VjdXJpdHk6c2VjdXJpdHlfdGVzdA==) on master @ [7552695eea44c6be9eecdd722b13c10821853a69](https://github.com/cockroachdb/cockroach/commits/7552695eea44c6be9eecdd722b13c10821853a69):


```
=== RUN   TestTLSCipherRestrict
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestTLSCipherRestrict2754239716
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:154: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    tls_test.go:251: -- test log scope end --
test logs left over in: outputs.zip/logTestTLSCipherRestrict2754239716
--- FAIL: TestTLSCipherRestrict (6.57s)
=== RUN   TestTLSCipherRestrict/no_cipher_set
    tls_test.go:200: expected wantError=false, got err=Get "https://127.0.0.1:39305/?cluster=test-tenant/_status/vars": net/http: request canceled (Client.Timeout exceeded while awaiting headers), resp=
--- FAIL: TestTLSCipherRestrict/no_cipher_set (2.00s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=2</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/security-engineering @cockroachdb/server @cockroachdb/product-security
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTLSCipherRestrict.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-49822

--- a/pkg/security/BUILD.bazel
+++ b/pkg/security/BUILD.bazel
@@ -93,6 +93,7 @@ go_test(
         "//pkg/server",
         "//pkg/testutils",
         "//pkg/testutils/serverutils",
+        "//pkg/testutils/skip",
         "//pkg/util/envutil",
         "//pkg/util/leaktest",
         "//pkg/util/log",

--- a/pkg/security/tls_test.go
+++ b/pkg/security/tls_test.go
@@ -22,6 +22,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/security/username"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/util/envutil"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -123,6 +124,8 @@ func verifyX509Cert(cert *x509.Certificate, dnsName string, roots *x509.CertPool
 func TestTLSCipherRestrict(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
+	skip.UnderStress(t, "http server accessing previous test's restriction fn")
+	skip.UnderRace(t)
 
 	// since the listener does not return rpc/sql/http connection errors, we
 	// need to have a separate hook to obtain and validate it.
@@ -142,6 +145,11 @@ func TestTLSCipherRestrict(t *testing.T) {
 	})()
 
 	ctx := context.Background()
+
+	// Start with a clean cipher configuration state
+	err := security.SetTLSCipherSuitesConfigured([]string{})
+	require.NoError(t, err)
+
 	s := serverutils.StartServerOnly(t, base.TestServerArgs{})
 	defer s.Stopper().Stop(ctx)
 	defer require.NoError(t, security.SetTLSCipherSuitesConfigured([]string{}))
@@ -181,8 +189,17 @@ func TestTLSCipherRestrict(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			// set the custom test ciphers
-			err := security.SetTLSCipherSuitesConfigured(tt.ciphers)
+			// First ensure we're starting with no restrictions
+			err := security.SetTLSCipherSuitesConfigured([]string{})
+			require.NoError(t, err)
+
+			// Reset the error container before each test
+			cipherErrC.Lock()
+			cipherErrC.err = nil
+			cipherErrC.Unlock()
+
+			// Now set the custom test ciphers
+			err = security.SetTLSCipherSuitesConfigured(tt.ciphers)
 			require.NoError(t, err)
 			// unset the ciphers after test
 			defer func() { _ = security.SetTLSCipherSuitesConfigured([]string{}) }()

